critiquebrainz/db/review.py

```diff
@@ -0,0 +1,608 @@
+        updates.append("license_id = :license_id")
+        updated_info["license_id"] = license_id
+
+    language = kwargs.pop("language", None)
+    if language is not None:
+        updates.append("language = :language")
+        updated_info["language"] = language
+
+    is_draft = kwargs.pop("is_draft", None)
+    if is_draft is not None:
+        if not drafted and is_draft: # If trying to convert published review into draft
+            raise BadRequest(lazy_gettext("Converting published reviews back to drafts is not allowed."))
+        updates.append("is_draft = :is_draft")
+        updated_info["is_draft"] = is_draft
+
+    text = kwargs.pop("text")
+    if kwargs:
+        raise TypeError('Unexpected **kwargs: %r' %kwargs)
+
+    setstr = ", ".join(updates)
+    query = sqlalchemy.text("""
+        UPDATE review
+           SET {}
+         WHERE id = :review_id
+    """.format(setstr))
+
+    if setstr:
+        updated_info["review_id"] = review_id
+        with db.engine.connect() as connection:
+            connection.execute(query, updated_info)
+    # Create new revision
+    new_revision = db_revision.create(review_id, text)
+    review = get_by_id(review_id)
+    cache.invalidate_namespace(REVIEW_CACHE_NAMESPACE)
+    return review
+
+
+def create(**kwargs):
+    if 'release_group' in kwargs:
+       entity_id = kwargs.pop('release_group')
+       entity_type = 'release_group'
+    else:
+        entity_id = kwargs.pop('entity_id')
+        entity_type = kwargs.pop('entity_type')
+
+    with db.engine.connect() as connection:
+        result = connection.execute(sqlalchemy.text("""
+            BEGIN;
+            INSERT INTO review
+            VALUES (:id, :entity_id, :entity_type, :user_id, :edits, :is_draft,
+            :is_hidden, :license_id, :language, :source, :source_url)
+         RETURNING id;
+       """), {
+           "id": str(uuid.uuid4()),
+           "entity_id": entity_id,
+           "entity_type": entity_type,
+           "user_id": kwargs.pop("user_id"),
+           "edits": 0,
+           "is_draft": kwargs.pop("is_draft", False),
+           "is_hidden": False,
+           "language": kwargs.pop("language", DEFAULT_LANG),
+           "license_id": kwargs.pop("license_id", DEFAULT_LICENSE_ID),
+           "source": kwargs.pop("source", None),
+           "source_url": kwargs.pop("source_url", None),
+        })
+        text = kwargs.pop("text")
+        if kwargs:
+            connection.execute(sqlalchemy.text("""
+                ROLLBACK;
+            """))
+            raise TypeError("Unexpected **kwargs: %r" % kwargs)
+        connection.execute(sqlalchemy.text("""
+            COMMIT;
+        """))
+        review_id = result.fetchone()[0]
+        review_revision = db_revision.create(review_id, text)
+        cache.invalidate_namespace(REVIEW_CACHE_NAMESPACE)
+    return get_by_id(review_id)
+
+
+def list_reviews(**kwargs):
```

Comment:
So the reason that all arguments were put into `**kwargs` is to force users of this function to explicitly specify them by keywords only. This was implemented before CB switched to Python 3. Python 3 has a better way of handling this: https://www.python.org/dev/peps/pep-3102/.